File: generated_code/cpp/solution_3555_3.cpp
#include <vector>
#include <queue>
#include <utility>

using namespace std;

class Solution {
public:
    vector<int> getFinalState(vector<int>& nums, int k, int multiplier) {
        // Custom comparator for the min-heap.
        // Elements are compared primarily by their current value.
        // If values are equal, the element with the smaller original index comes first.
        auto comp = [](const pair<int, int>& a, const pair<int, int>& b) {
            if (a.first != b.first) {
                return a.first > b.first; // Min-heap based on value
            }
            return a.second > b.second;   // Min-heap based on index for ties
        };
        
        // Priority queue stores pairs of {value, original_index}
        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(comp)> min_heap(comp);
        
        // Initialize the heap with all elements from nums
        for (int i = 0; i < nums.size(); ++i) {
            min_heap.push({nums[i], i});
        }
        
        // Perform k operations: always extract the minimum element,
        // multiply its value, and push it back into the heap.
        // Time Complexity per operation: O(log N)
        while (k > 0) {
            auto [val, idx] = min_heap.top();
            min_heap.pop();
            
            val *= multiplier;
            min_heap.push({val, idx});
            
            k--;
        }
        
        // Reconstruct the final state array using the original indices
        vector<int> result(nums.size());
        while (!min_heap.empty()) {
            auto [val, idx] = min_heap.top();
            min_heap.pop();
            result[idx] = val;
        }
        
        return result;
    }
};

File: generated_code/python/solution_3555_3.py
from typing import List
import heapq

class Solution:
    def getFinalState(self, nums: List[int], k: int, multiplier: int) -> List[int]:
        # If multiplier is 1, operations don't change the array values.
        if multiplier == 1:
            return nums
        
        # Initialize a min-heap with tuples of (value, original_index)
        # This naturally handles tracking the minimum value and breaking ties
        # by selecting the smallest index (first occurrence).
        heap = [(val, idx) for idx, val in enumerate(nums)]
        heapq.heapify(heap)
        
        # Perform the k operations
        for _ in range(k):
            val, idx = heapq.heappop(heap)
            new_val = val * multiplier
            nums[idx] = new_val
            heapq.heappush(heap, (new_val, idx))
            
        return nums